File: android/src/main/java/expo/modules/speechtotext/SpeechToTextModels.kt
package expo.modules.speechtotext

internal enum class SpeechToTextCleanupStatus(val rawValue: String) {
  Disabled("disabled"),
  Pending("pending"),
  Available("available"),
  ModelNotReady("model-not-ready"),
  UnsupportedDevice("unsupported-device"),
  UnsupportedLocale("unsupported-locale"),
  Unavailable("unavailable"),
  Failed("failed")
}

internal enum class SpeechToTextTranscriptionCapability(val rawValue: String) {
  Enhanced("enhanced"),
  Basic("basic"),
  Unavailable("unavailable")
}

internal enum class SpeechToTextModelAssetStatus(val rawValue: String) {
  Ready("ready"),
  Downloadable("downloadable"),
  Downloading("downloading"),
  Unavailable("unavailable")
}

internal data class SpeechToTextCapabilities(
  val transcription: SpeechToTextTranscriptionCapability,
  val cleanup: SpeechToTextCleanupStatus,
  val modelAssets: SpeechToTextModelAssetStatus,
  val onDeviceOnly: Boolean,
  val supportedLocale: String?
) {
  fun toMap(): Map<String, Any?> = mapOf(
    "transcription" to transcription.rawValue,
    "cleanup" to cleanup.rawValue,
    "modelAssets" to modelAssets.rawValue,
    "onDeviceOnly" to onDeviceOnly,
    "supportedLocale" to supportedLocale
  )
}

internal data class SpeechToTextTranscriptSegment(
  val text: String,
  val startMs: Double? = null,
  val endMs: Double? = null,
  val confidence: Double? = null
) {
  fun toMap(): Map<String, Any?> = mapOf(
    "text" to text,
    "startMs" to startMs,
    "endMs" to endMs,
    "confidence" to confidence
  )
}

File: ios/SpeechToTextModels.swift
import Foundation

enum SpeechToTextCleanupStatus: String {
  case disabled
  case pending
  case available
  case modelNotReady = "model-not-ready"
  case unsupportedDevice = "unsupported-device"
  case unsupportedLocale = "unsupported-locale"
  case unavailable
  case failed
}

enum SpeechToTextCleanupStyle: String {
  case dictation
  case note
  case message
}

enum SpeechToTextTranscriptionCapability: String {
  case enhanced
  case basic
  case unavailable
}

enum SpeechToTextModelAssetStatus: String {
  case ready
  case downloadable
  case downloading
  case unavailable
}

struct SpeechToTextCapabilities {
  let transcription: SpeechToTextTranscriptionCapability
  let cleanup: SpeechToTextCleanupStatus
  let modelAssets: SpeechToTextModelAssetStatus
  let supportedLocale: String?

  func asDictionary() -> [String: Any?] {
    [
      "transcription": transcription.rawValue,
      "cleanup": cleanup.rawValue,
      "modelAssets": modelAssets.rawValue,
      "onDeviceOnly": true,
      "supportedLocale": supportedLocale,
    ]
  }
}

struct SpeechToTextTranscriptSegment {
  let text: String
  let startMs: Double?
  let endMs: Double?
  let confidence: Double?

  func asDictionary() -> [String: Any?] {
    [
      "text": text,
      "startMs": startMs,
      "endMs": endMs,
      "confidence": confidence,
    ]
  }
}
